docs/python-grader/index.md

```diff
@@ -0,0 +1,203 @@
+class Test(PrairieLearnTestCase):
+
+## Plot grading enabled
+class Test(PrairieLearnTestCaseWithPlot):
+```
+
+These classes themselves extend `unittest.TestCase`, so any functionality from there is also available.
+
+Each test case is a separate method in the class, the names of these functions must be prefixed with `test_`, but any name that follows is arbitrary.  Test cases are run in sorted order by the `unittest` library, so one convention that can be used is to give tests numeric names to ensure their running order.
+
+Adding a name and point value to the test case is done by means of python decorators:
+`@points(val)` and `@name("name of the test case")`.  These will control the name of the case and the points awarded in the "Test Results" dropdown menu that are shown when a student submits a solution.  An example of a definition:
+
+```python
+@points(10)
+@name("Cool test case")
+def test_0(self):
+```
+
+Inside the test case implementation, the student answer variables and reference answer variables can be accessed as children of the tuples `self.st` and `self.ref`, respectively.  There are various helper functions to check correctness of different types of variables, these are defined in `code_feedback.py`.  These are taken from the RELATE grader, so this may be familiar to those with prior experience with RELATE.
+
+At the end of the test case, set the correctness of the answer using `feedback.set_percent()`.  This function takes a floating point number between 0 and 1 (inclusive), with 0 being completely *in*correct and 1 being completely correct.  By default, if no points are given the test case will be marked incorrect.
+
+The overall structure of a test case should look something like:
+
+```python
+from code_feedback import Feedback
+
+@points(10)
+@name("name of the test case")
+def test_0(self):
+   if Feedback.check_scalar('name of the variable', self.ref.variable_name, self.st.variable_names):
+       Feedback.set_percent(1)
+   else:
+       Feedback.set_percent(0)
+```
+
+Note that `Feedback.set_percent()` is used to set the _percentage_ correctness of the test case.  For example, if a test case is worth 10 points and `Feedback.set_percent(0.5)` is run, the student will be awarded 5 points.
+
+#### Multiple Iterations
+
+By setting the `total_iters` class variable, the test suite can be run for multiple iterations.  To prevent a specific test case from being run multiple times, you can add the `@not_repeated` decorator to it.
+
+#### Code Feedback
+
+The code feedback library contains built-in functions for checking correctness of various datatypes.  Here is a nonexhaustive list of them, for a more complete reference refer to the source file [on GitHub](https://github.com/PrairieLearn/PrairieLearn/blob/master/graders/python/python_autograder/code_feedback.py).  Note that all functions will perform some sort of sanity checking on user input and will not fail if, for example, the student does not define an input variable.
```

Comment:
Can we auto-generate proper docs from the code docstrings? See [this advice](https://docs.python-guide.org/writing/documentation/). I think Sphinx can [generate markdown](https://www.sphinx-doc.org/en/master/usage/markdown.html). We could either use the [RTD support](https://docs.readthedocs.io/en/stable/intro/getting-started-with-sphinx.html) or just run Sphinx ourselves in the [Makefile](https://github.com/PrairieLearn/PrairieLearn/blob/master/docs/Makefile) to make a Markdown (or HTML?) file, like we do for the diagram PNGs using graphviz.